File: python/sorting/topological_sort.py
"""
Topological Sort Algorithm (Kahn's Algorithm)

Description:
Topological sort is a linear ordering of vertices in a Directed Acyclic Graph (DAG)
such that for every directed edge u -> v, vertex u comes before v in the ordering.

This algorithm is a way of arranging tasks in an order so that every task is done
only after all the tasks it depends on are finished.

Time Complexity: O(V + E)
where V = number of vertices, E = number of edges

Space Complexity: O(V)

Note:
Topological sort is only possible for Directed Acyclic Graphs (DAGs).
If a cycle exists, topological sorting is not possible.
"""

from collections import deque
from typing import List

def topological_sort(V: int, adj: List[List[int]]) -> List[int]:
    """
    Performs topological sort on a directed graph
    :param V: number of vertices
    :param adj: adjacency list representing the graph
    :return: list containing topological order of vertices
    """
    indegree = [0] * V

    # Calculate indegree of each vertex
    for u in range(V):
        for v in adj[u]:
            indegree[v] += 1

    # Push all vertices with indegree 0 into queue
    q = deque()
    for i in range(V):
        if indegree[i] == 0:
            q.append(i)

    topo_order = []

    # Process vertices in BFS order
    while q:
        node = q.popleft()
        topo_order.append(node)

        # Reduce indegree of adjacent vertices
        for neighbor in adj[node]:
            indegree[neighbor] -= 1
            if indegree[neighbor] == 0:
                q.append(neighbor)

    # Cycle detection
    if len(topo_order) != V:
        print("Error: Graph contains a cycle.")
        return []

    return topo_order

def print_vector(arr: List[int]) -> None:
    """
    Prints a list
    :param arr: list to print
    """
    print(arr)

# Example usage and test cases
if __name__ == "__main__":
    print("=== Topological Sort Examples ===\n")

    # Test Case 1: Simple DAG
    V1 = 6
    adj1 = [[] for _ in range(V1)]

    adj1[5].append(2)
    adj1[5].append(0)
    adj1[4].append(0)
    adj1[4].append(1)
    adj1[2].append(3)
    adj1[3].append(1)

    print("Test 1 - Simple DAG:")
    result1 = topological_sort(V1, adj1)
    print("Topological Order:", end=" ")
    print_vector(result1)
    print()

    # Test Case 2: Linear dependency
    V2 = 4
    adj2 = [[] for _ in range(V2)]

    adj2[0].append(1)
    adj2[1].append(2)
    adj2[2].append(3)

    print("Test 2 - Linear dependency:")
    result2 = topological_sort(V2, adj2)
    print("Topological Order:", end=" ")
    print_vector(result2)
    print()

    # Test Case 3: Single node
    V3 = 1
    adj3 = [[] for _ in range(V3)]

    print("Test 3 - Single node:")
    result3 = topological_sort(V3, adj3)
    print("Topological Order:", end=" ")
    print_vector(result3)
    print()

    # Test Case 4: Disconnected DAG
    V4 = 5
    adj4 = [[] for _ in range(V4)]

    adj4[0].append(1)
    adj4[2].append(3)

    print("Test 4 - Disconnected DAG:")
    result4 = topological_sort(V4, adj4)
    print("Topological Order:", end=" ")
    print_vector(result4)
    print()

    # Test Case 5: Graph with a cycle
    V5 = 3
    adj5 = [[] for _ in range(V5)]

    adj5[0].append(1)
    adj5[1].append(2)
    adj5[2].append(0)

    print("Test 5 - Graph with a cycle:")
    result5 = topological_sort(V5, adj5)
    print("Topological Order:", end=" ")
    print_vector(result5)
    print()

File: c/sorting/topological_sort.c
/*
 * Topological Sort Algorithm (Kahn's Algorithm)
 *
 * Description:
 * Topological sort is a linear ordering of vertices in a Directed Acyclic Graph (DAG)
 * such that for every directed edge u -> v, vertex u comes before v in the ordering.
 * 
 * This algorithm is a way of arranging tasks in an order so that every task is done only after all the tasks it depends on are finished.
 * 
 * Time Complexity: O(V + E)
 * where V = number of vertices, E = number of edges
 *
 * Space Complexity: O(V)
 *
 * Note:
 * Topological sort is only possible for Directed Acyclic Graphs (DAGs)
 * If a cycle exists, topological sorting is not possible
 */

#include <stdio.h>

const int max_vertices = 10;

/**
 * Performs topological sort on a directed graph
 *
 * @param graph: adjacency matrix of the graph
 * @param vertices: number of vertices in the graph
 */
void topological_sort(int graph[max_vertices][max_vertices], int vertices) {
    int indegree[max_vertices] = {0};
    int queue[max_vertices];
    int front = 0, rear = 0;
    int count = 0;

    // Calculate in-degree of each vertex
    for (int i = 0; i < vertices; i++) {
        for (int j = 0; j < vertices; j++) {
            if (graph[i][j] == 1) {
                indegree[j]++;
            }
        }
    }

    // Add vertices with in-degree 0 to the queue
    for (int i = 0; i < vertices; i++) {
        if (indegree[i] == 0) {
            queue[rear++] = i;
        }
    }

    printf("Topological Order: ");

    // Process the queue
    while (front < rear) {
        int current = queue[front++];
        printf("%d ", current);
        count++;

        // Reduce in-degree of adjacent vertices
        for (int i = 0; i < vertices; i++) {
            if (graph[current][i] == 1) {
                indegree[i]--;
                if (indegree[i] == 0) {
                    queue[rear++] = i;
                }
            }
        }
    }

    // Check for cycle
    if (count != vertices) {
        printf("\nError: Graph contains a cycle.\n");
    } else {
        printf("\n");
    }
}

/**
 * Prints the adjacency matrix of the graph
 *
 * @param graph: adjacency matrix
 * @param vertices: number of vertices
 */
void print_graph(int graph[max_vertices][max_vertices], int vertices) {
    printf("Adjacency Matrix:\n");
    for (int i = 0; i < vertices; i++) {
        for (int j = 0; j < vertices; j++) {
            printf("%d ", graph[i][j]);
        }
        printf("\n");
    }
    printf("\n");
}

// Example usage and test cases
int main() {
    printf("=== Topological Sort Examples ===\n\n");

    // Test Case 1: Valid DAG
    int graph1[max_vertices][max_vertices] = {
        {0, 1, 1, 0, 0},
        {0, 0, 0, 1, 0},
        {0, 0, 0, 1, 0},
        {0, 0, 0, 0, 1},
        {0, 0, 0, 0, 0}
    };
    int vertices1 = 5;

    printf("Test 1 - Directed Acyclic Graph:\n");
    print_graph(graph1, vertices1);
    topological_sort(graph1, vertices1);
    printf("\n");

    // Test Case 2: Graph with a cycle
    int graph2[max_vertices][max_vertices] = {
        {0, 1, 0},
        {0, 0, 1},
        {1, 0, 0}
    };
    int vertices2 = 3;

    printf("Test 2 - Graph with Cycle:\n");
    print_graph(graph2, vertices2);
    topological_sort(graph2, vertices2);
    printf("\n");

    // Test Case 3: Single vertex
    int graph3[max_vertices][max_vertices] = {
        {0}
    };
    int vertices3 = 1;

    printf("Test 3 - Single Vertex:\n");
    print_graph(graph3, vertices3);
    topological_sort(graph3, vertices3);
    printf("\n");

    return 0;
}